diff --git a/include/nodes/internal/NodePainterDelegate.hpp b/include/nodes/internal/NodePainterDelegate.hpp
index d532619..f22496c 100644
--- a/include/nodes/internal/NodePainterDelegate.hpp
+++ b/include/nodes/internal/NodePainterDelegate.hpp
@@ -1,6 +1,6 @@
 #pragma once
 
-#include <QPainter>
+#include <QtGui/QPainter>
 
 #include "NodeGeometry.hpp"
 #include "NodeDataModel.hpp"
diff --git a/include/nodes/internal/QStringStdHash.hpp b/include/nodes/internal/QStringStdHash.hpp
index 3d47a3f..cc5eb68 100644
--- a/include/nodes/internal/QStringStdHash.hpp
+++ b/include/nodes/internal/QStringStdHash.hpp
@@ -1,5 +1,11 @@
 #pragma once
 
+#include <QtCore/QtGlobal>
+
+#if (QT_VERSION < QT_VERSION_CHECK(5, 14, 0))
+
+// As of 5.14 there is a specialization std::hash<QString>
+
 #include <functional>
 
 #include <QtCore/QString>
@@ -17,3 +23,5 @@ struct hash<QString>
   }
 };
 }
+
+#endif
